drivers/sensor/bmi08x/bmi08x_accel.c

```diff
@@ -0,0 +1,775 @@
+
+	if (odr < BMI08X_ACCEL_ODR_12_5_HZ) {
+		return odr;
+	}
+
+	return bmi08x_accel_reg_field_update(dev, BMI08X_REG_ACCEL_CONF, 0, BMI08X_ACCEL_ODR_MASK,
+					     (uint8_t)odr);
+}
+
+static const struct bmi08x_range bmi085_acc_range_map[] = {
+	{2, BMI085_ACCEL_RANGE_2G},
+	{4, BMI085_ACCEL_RANGE_4G},
+	{8, BMI085_ACCEL_RANGE_8G},
+	{16, BMI085_ACCEL_RANGE_16G},
+};
+#define BMI085_ACC_RANGE_MAP_SIZE ARRAY_SIZE(bmi085_acc_range_map)
+
+static const struct bmi08x_range bmi088_acc_range_map[] = {
+	{3, BMI088_ACCEL_RANGE_3G},
+	{6, BMI088_ACCEL_RANGE_6G},
+	{12, BMI088_ACCEL_RANGE_12G},
+	{24, BMI088_ACCEL_RANGE_24G},
+};
+#define BMI088_ACC_RANGE_MAP_SIZE ARRAY_SIZE(bmi088_acc_range_map)
+
+static int bmi08x_acc_range_set(const struct device *dev, int32_t range)
+{
+	struct bmi08x_accel_data *data = dev->data;
+	int32_t reg_val = -1;
+
+	if (data->accel_chip_id == BMI085_ACCEL_CHIP_ID) {
+		reg_val = bmi08x_range_to_reg_val(range, bmi085_acc_range_map,
+						  BMI085_ACC_RANGE_MAP_SIZE);
+	} else if (data->accel_chip_id == BMI088_ACCEL_CHIP_ID) {
+		reg_val = bmi08x_range_to_reg_val(range, bmi088_acc_range_map,
+						  BMI088_ACC_RANGE_MAP_SIZE);
+	} else {
+		return -ENXIO;
+	}
+
+	if (reg_val < 0) {
+		return reg_val;
+	}
+
+	if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_RANGE, reg_val & 0xff) < 0) {
+		return -EIO;
+	}
+
+	data->scale = BMI08X_ACC_SCALE(range);
+
+	return 0;
+}
+
+static int bmi08x_acc_config(const struct device *dev, enum sensor_channel chan,
+			     enum sensor_attribute attr, const struct sensor_value *val)
+{
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		return bmi08x_acc_range_set(dev, sensor_ms2_to_g(val));
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		return bmi08x_acc_odr_set(dev, val->val1, val->val2 / 1000);
+	default:
+		LOG_ERR("Accel attribute not supported.");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int bmi08x_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr, const struct sensor_value *val)
+{
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return -EBUSY;
+	}
+#endif
+
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		return bmi08x_acc_config(dev, chan, attr, val);
+	default:
+		LOG_ERR("attr_set() not supported on this channel.");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int bmi08x_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct bmi08x_accel_data *data = dev->data;
+	size_t i;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return -EBUSY;
+	}
+#endif
+
+	pm_device_busy_set(dev);
+
+	if (bmi08x_accel_transceive(dev, BMI08X_REG_ACCEL_X_LSB | (1 << 7), false, data->acc_sample,
```

Comment:
BMI08X_REG_ACCEL_X_LSB | (1 << 7)

ACC_X_LSB address is 0x12, I do not know why (0x12 | (1 << 7)) = 0x92?

Reply:
In transceive_spi you can add reg address with  (uint8_t addr = reg & 0x7f for write uint8_t addr = reg | 0x80 for read.)

Reply:
I should be using the BIT(7) macro here, but that bit is used to indicate a read

Reply:
In datasheet
```
6.1.2 SPI interface of accelerometer part
The data bits are used as follows: 
• Bit #0: Read/Write bit. When 0, the data SDI is written into the chip. When 1, the data SDO from 
the chip is read. 
• Bit #1-7: Address AD(6:0). 
• Bit #8-15:  
o When in write mode, these are the data SDI, which will be written into the address. 
o When in read mode, these bits contain unpredictable values, and the user has to read 
Bit #16-23 to get the actual data from the reading address.  
```
that means only in spi mode read/write is using the register address Bit0, 
```
6.2 Inter-Integrated Circuit (I²C)
I²C write access and I²C read access
```
in I2C mode the read/write is in slave address bit7

the operate in register address or slave address will be set in low level api, not here.

Reply:
if you read the rest of the code, and further into the zephyr i2c api, you'll see that r/w bit is handled there...

that "bit 0" (using the datasheet terminology) of the reg address is just a dummy in I2C, so if bit 0 is still set, it doesn't matter

Reply:
> if you read the rest of the code, and further into the zephyr i2c api, you'll see that r/w bit is handled there...
> 
> that "bit 0" (using the datasheet terminology) of the reg address is just a dummy in I2C, so if bit 0 is still set, it doesn't matter

I am using bmi088 in I2C mode, in this function`if (bmi08x_accel_transceive(dev, BMI08X_REG_ACCEL_X_LSB | (1 << 7), false, data->acc_sample,`   it read the register address is 0x92 not 0x12, so I could not get right value

Reply:
What value are you getting? That MSb is a dummy bit.

Are you getting the correct chip id as well? I'd be reading register 0x80 for that.

Reply:
I see. you have changed it, this commit can be closed
```
	if (bmi08x_accel_read(dev, BMI08X_REG_ACCEL_X_LSB, (uint8_t *)data->acc_sample,
			      sizeof(data->acc_sample)) < 0) {
		pm_device_busy_clear(dev);
		return -EIO;
	}
```

and
```
	if (bmi08x_gyro_read(dev, BMI08X_REG_GYRO_X_LSB, (uint8_t *)bmi08x->gyr_sample,
			     sizeof(bmi08x->gyr_sample)) < 0) {
		return -EIO;
	}
```